pkg/util/docker/docker_util.go

```diff
@@ -204,59 +212,11 @@ func (d *DockerUtil) ResolveImageName(image string) (string, error) {
 // It is similar to ResolveImageName except it tries to match the image to the container Config.Image.
 // For non-sha names we will just return the name as-is.
 func (d *DockerUtil) ResolveImageNameFromContainer(co types.ContainerJSON) (string, error) {
-	image := co.Image
-	if !strings.Contains(image, "sha256:") {
-		return image, nil
+	if co.Config.Image != "" && !isImageShaID(co.Config.Image) && !isImageRepoDigest(co.Config.Image) {
+		return co.Config.Image, nil
 	}
 
-	d.Lock()
-	defer d.Unlock()
-	if _, ok := d.imageNameBySha[image]; !ok {
-		ctx, cancel := context.WithTimeout(context.Background(), d.queryTimeout)
-		defer cancel()
-		r, _, err := d.cli.ImageInspectWithRaw(ctx, image)
-		if err != nil {
-			// Only log errors that aren't "not found" because some images may
-			// just not be available in docker inspect.
-			if !client.IsErrNotFound(err) {
-				return image, err
-			}
-			d.imageNameBySha[image] = image
-		}
-
-		imageName := getBestImageName(r, co.Config.Image)
-		if imageName != "" {
-			d.imageNameBySha[image] = imageName
-		}
-	}
-	return d.imageNameBySha[image], nil
-}
-
-func getBestImageName(r types.ImageInspect, configImage string) string {
-	var imageName string
-	// Try RepoTags first and fall back to RepoDigest otherwise.
-	if len(r.RepoTags) == 1 {
-		imageName = r.RepoTags[0]
-	} else if len(r.RepoTags) > 1 {
-		// If one of the RepoTags is the tag used to run the image, then set that
-		// as the image tag. Otherwise, use the first one (random)
-		for _, t := range r.RepoTags {
```

Comment:
i'm trying to think about whether there are any downsides to returning the config.Image in the case that the corresponding RepoTag is gone. as far as i can tell that's the main change in removing this method. can you think of any?

Reply:
I'm leaning towards thinking it's a positive change. If we have a metric tracking a `config.Image` and the corresponding repotag got deleted, that metric can potentially lose the expected image tag once the cache expiry happens (10min). This new behavior favors what is configured in the container (how the container was configured when started) rather than the state of the image id and it's related repotags.

Reply:
i agree! i'm just nervous about any downsides we may not have foreseen. one thing i can think of is that this could change the existing image tag for some users (if config.Image is not present as a repoTag).

Reply:
Right, so that's the case where a metric could be tracking a deleted tag, from the example in this PR - if a container started was reporting `redis:bar` since it's the first repotag for that ID, but with this change it will now report `redis:latest`.